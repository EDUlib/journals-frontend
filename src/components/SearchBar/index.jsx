import React from 'react';
import { SearchField } from '@edx/paragon';
import qs from 'query-string';
import PropTypes from 'prop-types';

const SEARCH_PATH = '/search/';

class SearchBar extends React.Component {
  constructor(props) {
    super(props);
    this.search = this.search.bind(this);
  }

  search(query) {
    const MATCH_PHRASE_START_CHAR = '"';
    const MATCH_PHRASE_END_CHAR = '"';
    let searchOperator = 'or';
    if (query.startsWith(MATCH_PHRASE_START_CHAR) && query.endsWith(MATCH_PHRASE_END_CHAR)) {
      searchOperator = 'and';
    }
    const cleanQuery = query.replace(new RegExp(MATCH_PHRASE_START_CHAR, 'g'), '');
    const options = {
      query: cleanQuery,
      operator: searchOperator,
      journalId: this.props.journalId,
    };
    this.props.history.push(`${SEARCH_PATH}?${qs.stringify(options)}`);
  }

  render() {
    return <SearchField onSubmit={(value) => { this.search(value); }} />;
  }
}

SearchBar.defaultProps = {
  journalId: 0,
};

SearchBar.propTypes = {
  journalId: PropTypes.number,
  history: PropTypes.shape({
    push: PropTypes.func,
  }).isRequired,
};

export default SearchBar;
