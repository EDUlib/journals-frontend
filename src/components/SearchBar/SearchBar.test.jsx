import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import SearchBar from './index';

import history from '../../data/history';

const props = {
  history,
};

function search(input, searchQuery) {
  input.simulate('change', { target: { value: searchQuery } });
  input.simulate('keypress', { key: 'Enter' });
}

describe('<SearchBar />', () => {
  describe('renders', () => {
    it('correctly', () => {
      const tree = renderer
        .create(<SearchBar {...props} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('search', () => {
    const wrapper = mount(<SearchBar {...props} />);
    const input = wrapper.find('input');
    const searchQuery = 'foo';

    it('updates input', () => {
      expect(wrapper.find('input').prop('value')).toBe('');
      search(input, searchQuery);
      expect(wrapper.find('input').prop('value')).toBe(searchQuery);
    });

    it('with null journal', () => {
      wrapper.setProps({ journalId: null });
      expect(wrapper.find('input').prop('value')).toBe(searchQuery);
      const newQuery = 'bar';
      search(input, newQuery);
      expect(wrapper.find('input').prop('value')).toBe(newQuery);
    });

    it('with OR operator', () => {
      search(input, searchQuery);
      expect(wrapper.prop('history').location.search).toMatch(`query=${searchQuery}`);
      expect(wrapper.prop('history').location.search).toMatch('operator=or');
    });

    it('with AND operator', () => {
      search(input, `"${searchQuery}"`);
      expect(wrapper.prop('history').location.search).toMatch(`query=${searchQuery}`);
      expect(wrapper.prop('history').location.search).toMatch('operator=and');
    });
  });
});
