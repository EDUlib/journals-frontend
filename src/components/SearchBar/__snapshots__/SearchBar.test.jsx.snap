// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> renders correctly 1`] = `
<div
  className="border search-field"
  onBlur={[Function]}
  onFocus={[Function]}
>
  <div
    className="form-group form-inline"
  >
    <label
      className=""
      htmlFor="asInput4"
      id="label-asInput4"
    >
      Search:
    </label>
    <div
      className="input-group"
    >
      <div
        className="input-group-prepend"
      />
      <input
        aria-describedby="error-asInput4"
        aria-invalid={false}
        autoComplete="off"
        className="form-control is-invalid-nodanger input no-clear-btn"
        disabled={false}
        id="asInput4"
        name="search"
        onBlur={[Function]}
        onChange={[Function]}
        onKeyPress={[Function]}
        placeholder=""
        readOnly={false}
        required={false}
        themes={Array []}
        type="search"
        value=""
      />
      <div
        className="input-group-append"
      >
        <button
          className="btn search-btn"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <span
            aria-hidden={true}
            className="fa fa-search"
            id="Icon1"
          />
          <span
            className="sr-only"
          >
            Submit search
          </span>
        </button>
      </div>
    </div>
    <div
      aria-live="polite"
      className="invalid-feedback invalid-feedback-nodanger"
      id="error-asInput4"
    >
      <span />
    </div>
  </div>
</div>
`;
